# Import packages

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

# Read data

Supply Chain Shipment Pricing Dataset from USAID, https://data.usaid.gov/HIV-AIDS/Supply-Chain-Shipment-Pricing-Dataset/a3rc-nmf6/about_data, consisting of information about Antiretroviral (ARV) and HIV Lab Shipments to Supported Countries.

In [2]:
path = "C:/Users/Bertha/Documents/SJSU/5_Spring2024/DATA_240-Data_Mining_Analytics/DATA240_Project/DATA240/data/Supply_Chain_Shipment_Pricing_Dataset_20240302.csv"
data = pd.read_csv(path, index_col="id") # ID col is treated as the index

Some IDs are missing in the dataset such as id=2,5,6-14, etc.

In [3]:
data

project code            pq # po / so # asn/dn #        country  \
id                                                                     
1       100-CI-T01  Pre-PQ Process    SCMS-4    ASN-8  Côte d'Ivoire   
3       108-VN-T01  Pre-PQ Process   SCMS-13   ASN-85        Vietnam   
4       100-CI-T01  Pre-PQ Process   SCMS-20   ASN-14  Côte d'Ivoire   
15      108-VN-T01  Pre-PQ Process   SCMS-78   ASN-50        Vietnam   
16      108-VN-T01  Pre-PQ Process   SCMS-81   ASN-55        Vietnam   
...            ...             ...       ...      ...            ...   
86818   103-ZW-T30       FPQ-15197  SO-50020  DN-4307       Zimbabwe   
86819   104-CI-T30       FPQ-15259  SO-50102  DN-4313  Côte d'Ivoire   
86821   110-ZM-T30       FPQ-14784  SO-49600  DN-4316         Zambia   
86822   200-ZW-T30       FPQ-16523  SO-51680  DN-4334       Zimbabwe   
86823   103-ZW-T30       FPQ-15197  SO-50022  DN-4336       Zimbabwe   

      managed by  fulfill via vendor inco term shipment mode  \
id                                                             
1       PMO - US  Direct Drop              EXW           Air   
3       PMO - US  Direct Drop              EXW           Air   
4       PMO - US  Direct Drop              FCA           Air   
15      PMO - US  Direct Drop              EXW           Air   
16      PMO - US  Direct Drop              EXW           Air   
...          ...          ...              ...           ...   
86818   PMO - US     From RDC   N/A - From RDC         Truck   
86819   PMO - US     From RDC   N/A - From RDC         Truck   
86821   PMO - US     From RDC   N/A - From RDC         Truck   
86822   PMO - US     From RDC   N/A - From RDC         Truck   
86823   PMO - US     From RDC   N/A - From RDC         Truck   

      pq first sent to client date  ... unit of measure (per pack)  \
id                                  ...                              
1                   Pre-PQ Process  ...                         30   
3                   Pre-PQ Process  ...                        240   
4                   Pre-PQ Process  ...                        100   
15                  Pre-PQ Process  ...                         60   
16                  Pre-PQ Process  ...                         60   
...                            ...  ...                        ...   
86818                   10/16/2014  ...                         60   
86819                   10/24/2014  ...                         60   
86821                    8/12/2014  ...                         30   
86822                     7/1/2015  ...                         60   
86823                   10/16/2014  ...                         60   

      line item quantity line item value pack price unit price  \
id                                                               
1                     19          551.00      29.00       0.97   
3                   1000         6200.00       6.20       0.03   
4                    500        40000.00      80.00       0.80   
15                 31920       127360.80       3.99       0.07   
16                 38000       121600.00       3.20       0.05   
...                  ...             ...        ...        ...   
86818             166571       599655.60       3.60       0.06   
86819              21072       137389.44       6.52       0.11   
86821             514526      5140114.74       9.99       0.33   
86822              17465       113871.80       6.52       0.11   
86823              36639        72911.61       1.99       0.03   

                   manufacturing site first line designation  \
id                                                             
1          Ranbaxy Fine Chemicals LTD                   True   
3           Aurobindo Unit III, India                   True   
4       ABBVIE GmbH & Co.KG Wiesbaden                   True   
15      Ranbaxy, Paonta Shahib, India                   True   
16          Aurobindo Unit III, India                   True   
...               

# Data Cleaning

In [4]:
# for each col replace special characters
newcols = [i.strip().replace(" ","_").replace("#","num").replace("/","_").replace("(","").replace(")","") for i in data.columns]
data.columns = newcols

In [5]:
data.shape # 10324 rows and 32 cols

(10324, 32)

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10324 entries, 1 to 86823
Data columns (total 32 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   project_code                  10324 non-null  object 
 1   pq_num                        10324 non-null  object 
 2   po___so_num                   10324 non-null  object 
 3   asn_dn_num                    10324 non-null  object 
 4   country                       10324 non-null  object 
 5   managed_by                    10324 non-null  object 
 6   fulfill_via                   10324 non-null  object 
 7   vendor_inco_term              10324 non-null  object 
 8   shipment_mode                 9964 non-null   object 
 9   pq_first_sent_to_client_date  10324 non-null  object 
 10  po_sent_to_vendor_date        10324 non-null  object 
 11  scheduled_delivery_date       10324 non-null  object 
 12  delivered_to_client_date      10324 non-null  object 
 13  delive

In [7]:
# 3 cols out of 32 total cols have null values
data.isnull().sum()

project_code                       0
pq_num                             0
po___so_num                        0
asn_dn_num                         0
country                            0
managed_by                         0
fulfill_via                        0
vendor_inco_term                   0
shipment_mode                    360
pq_first_sent_to_client_date       0
po_sent_to_vendor_date             0
scheduled_delivery_date            0
delivered_to_client_date           0
delivery_recorded_date             0
product_group                      0
sub_classification                 0
vendor                             0
item_description                   0
molecule_test_type                 0
brand                              0
dosage                          1736
dosage_form                        0
unit_of_measure_per_pack           0
line_item_quantity                 0
line_item_value                    0
pack_price                         0
unit_price                         0
m

In [8]:
# check duplicates
data.duplicated().sum()

4

In [9]:
# print the 4 duplicates (8 total rows) to the screen
data[data.duplicated(keep=False)]

project_code          pq_num  po___so_num asn_dn_num      country  \
id                                                                        
10692   125-HT-T01  Pre-PQ Process      SO-1291     DN-686        Haiti   
11246   125-HT-T01  Pre-PQ Process      SO-1291     DN-686        Haiti   
11882   100-ZW-T01  Pre-PQ Process       SO-716     DN-770     Zimbabwe   
12146   100-ZW-T01  Pre-PQ Process       SO-710     DN-488     Zimbabwe   
12147   100-ZW-T01  Pre-PQ Process       SO-710     DN-488     Zimbabwe   
12792   100-ZW-T01  Pre-PQ Process       SO-716     DN-770     Zimbabwe   
43250   105-SS-T30       FPQ-12623  SCMS-200920  ASN-21751  South Sudan   
68128   105-SS-T30       FPQ-12623  SCMS-200920  ASN-21751  South Sudan   

      managed_by  fulfill_via vendor_inco_term shipment_mode  \
id                                                             
10692   PMO - US     From RDC   N/A - From RDC           Air   
11246   PMO - US     From RDC   N/A - From RDC           Air   
11882   PMO - US     From RDC   N/A - From RDC           Air   
12146   PMO - US     From RDC   N/A - From RDC           Air   
12147   PMO - US     From RDC   N/A - From RDC           Air   
12792   PMO - US     From RDC   N/A - From RDC           Air   
43250   PMO - US  Direct Drop              EXW           Air   
68128   PMO - US  Direct Drop              EXW           Air   

      pq_first_sent_to_client_date  ... unit_of_measure_per_pack  \
id                                  ...                            
10692               Pre-PQ Process  ...                      100   
11246               Pre-PQ Process  ...                      100   
11882               Pre-PQ Process  ...                       30   
12146               Pre-PQ Process  ...                       30   
12147               Pre-PQ Process  ...                       30   
12792               Pre-PQ Process  ...                       30   
43250                    7/18/2013  ...                       20   
68128                    7/18/2013  ...                       20   

      line_item_quantity line_item_value pack_price unit_price  \
id                                                               
10692                 10           720.0      72.00       0.72   
11246                 10           720.0      72.00       0.72   
11882               1860         23101.2      12.42       0.41   
12146               1860         23101.2      12.42       0.41   
12147               1860         23101.2      12.42       0.41   
12792               1860         23101.2      12.42       0.41   
43250                 25           800.0      32.00       1.60   
68128                 25           800.0      32.00       1.60   

              manufacturing_site first_line_designation  \
id                                                        
10692            Inverness Japan                  False   
11246            Inverness Japan                  False   
11882  Aurobindo Unit III, India                  False   
12146          Cipla, Goa, India                  False   
12147          Cipla, Goa, India                  False   
12792  Aurobindo Unit III, India                  False   
43250       Trinity Biotech, Plc                  False   
68128       Trinity Biotech, Plc                  False   

                weight_kilograms           freight_cost_usd  \
id                                                            
10692     See DN-686 (ID#:10691)     See DN-686 (ID#:10691)   
11246     See DN-686 (ID#:10691)     See DN-686 (ID#:10691)   
11882     See DN-770 (ID#:10959)     See DN-770 (ID#:10959)   
12146     See DN-488 (ID#:10652)     See DN-488 (ID#:10652)   
12147     See DN-488 (ID#:10652)     See DN-488 (ID#:10652)   
12792     See DN-770 (ID#:10959)     See DN-770 (ID#:10959)   
43250  See ASN-21751 (ID#:13968)  See ASN-21751 (ID#:13968)   
68128  See ASN-21751 (ID#:13968)  See ASN-21751 (ID#:13968)   

      line_item_insurance_usd  
id                

These rows are exactly identical, drop 4 rows. Likely a human error

In [10]:
# drop duplicate rows
data = data.drop_duplicates()

In [11]:
# check dataframe
data[data.duplicated(keep=False)]

Empty DataFrame
Columns: [project_code, pq_num, po___so_num, asn_dn_num, country, managed_by, fulfill_via, vendor_inco_term, shipment_mode, pq_first_sent_to_client_date, po_sent_to_vendor_date, scheduled_delivery_date, delivered_to_client_date, delivery_recorded_date, product_group, sub_classification, vendor, item_description, molecule_test_type, brand, dosage, dosage_form, unit_of_measure_per_pack, line_item_quantity, line_item_value, pack_price, unit_price, manufacturing_site, first_line_designation, weight_kilograms, freight_cost_usd, line_item_insurance_usd]
Index: []

[0 rows x 32 columns]

In [12]:
# after dropping duplicates, check 
data.duplicated().sum()

0

In [13]:
data.shape # Before: 10324 rows
           # After: 10320 rows

(10320, 32)

# Convert data types

## Explain why `po_sent_to_vendor_date`  and `pq_first_sent_to_client_date` has missing data values

PQ is the price quote, defined as a document that lists an approximately fixed price of goods. It's an informal statement between the buyer and supplier that states an approximate cost of the goods or services in the exchange. In this case, a quote or approximate cost for a bulk of pharmaceutical product to be exchanged between the suppliers (vendor column) and the buyers.

On the other hand, the PO is the purchase order that is an official document that lists the exact prices and details of the purchase. It's a legal document the buyer uses to send to the supplier when making an order, listing the item, description, unit price, total price, quantity, etc. 


Sometimes the date the price quote was exchanged between the vendor and buyer was not captured, likely due to lack of communication via email, phone, etc. Many of these exchanges between the supplier sales person and a buyer may be exchanged verbally so an informal document listing the price may not always be captured, which makes sense in the realworld

Actually it's a common problem for sales representatives as they spend a lot of their time manually inputting or transferring data, leading to inconsistent document formatting, according to https://www.crmbuyer.com/story/how-sales-reps-can-solve-3-nagging-document-problems-86134.html. 

Customer Relationship Management (CRM) Tools are used to input customer and sales data to help businesspeople track communication and information relating to their sales. Although the purpose of a CRM is to make it easier and more efficient in dealing with multiple customers and leads/deals at the same time, in reality sales people end up with the bulk of the manual work for these CRM Tools. Instead of talking to prospectual customers, they spend a bulk of their time trying to keep all their information up-to-date, which is a big problem tradition CRM systems still have and impact sales representatives.

In [14]:
data['po_sent_to_vendor_date'] = data['po_sent_to_vendor_date'].replace(['Date Not Captured', 'N/A - From RDC'],np.nan)
data['pq_first_sent_to_client_date'] = data['po_sent_to_vendor_date'].replace(['Date Not Captured', 'Pre-PQ Process'],np.nan)

C:\Users\Bertha\AppData\Local\Temp\ipykernel_6740\70630604.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['po_sent_to_vendor_date'] = data['po_sent_to_vendor_date'].replace(['Date Not Captured', 'N/A - From RDC'],np.nan)
C:\Users\Bertha\AppData\Local\Temp\ipykernel_6740\70630604.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['pq_first_sent_to_client_date'] = data['po_sent_to_vendor_date'].replace(['Date Not Captured', 'Pre-PQ Process'],np.nan)


In [15]:
date_cols = ['po_sent_to_vendor_date',
 'scheduled_delivery_date',
 'delivered_to_client_date',
 'delivery_recorded_date',
 'pq_first_sent_to_client_date'
 ]
display(data[date_cols])
for d in date_cols:
    print(d)
    data[d] = pd.to_datetime(data[d])

po_sent_to_vendor_date scheduled_delivery_date delivered_to_client_date  \
id                                                                              
1                        NaN                2-Jun-06                 2-Jun-06   
3                        NaN               14-Nov-06                14-Nov-06   
4                        NaN               27-Aug-06                27-Aug-06   
15                       NaN                1-Sep-06                 1-Sep-06   
16                       NaN               11-Aug-06                11-Aug-06   
...                      ...                     ...                      ...   
86818                    NaN               31-Jul-15                15-Jul-15   
86819                    NaN               31-Jul-15                 6-Aug-15   
86821                    NaN               31-Aug-15                25-Aug-15   
86822                    NaN                9-Sep-15                 4-Aug-15   
86823                    NaN               31-Aug-15                 4-Aug-15   

      delivery_recorded_date pq_first_sent_to_client_date  
id                                                         
1                   2-Jun-06                          NaN  
3                  14-Nov-06                          NaN  
4                  27-Aug-06                          NaN  
15                  1-Sep-06                          NaN  
16                 11-Aug-06                          NaN  
...                      ...                          ...  
86818              20-Jul-15                          NaN  
86819               7-Aug-15                          NaN  
86821               3-Sep-15                          NaN  
86822              11-Aug-15                          NaN  
86823              11-Aug-15                          NaN  

[10320 rows x 5 columns]

C:\Users\Bertha\AppData\Local\Temp\ipykernel_6740\2910687795.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[d] = pd.to_datetime(data[d])
C:\Users\Bertha\AppData\Local\Temp\ipykernel_6740\2910687795.py:10: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  data[d] = pd.to_datetime(data[d])


po_sent_to_vendor_date
scheduled_delivery_date


C:\Users\Bertha\AppData\Local\Temp\ipykernel_6740\2910687795.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[d] = pd.to_datetime(data[d])
C:\Users\Bertha\AppData\Local\Temp\ipykernel_6740\2910687795.py:10: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  data[d] = pd.to_datetime(data[d])


delivered_to_client_date


C:\Users\Bertha\AppData\Local\Temp\ipykernel_6740\2910687795.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[d] = pd.to_datetime(data[d])
C:\Users\Bertha\AppData\Local\Temp\ipykernel_6740\2910687795.py:10: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  data[d] = pd.to_datetime(data[d])


delivery_recorded_date
pq_first_sent_to_client_date


C:\Users\Bertha\AppData\Local\Temp\ipykernel_6740\2910687795.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[d] = pd.to_datetime(data[d])
C:\Users\Bertha\AppData\Local\Temp\ipykernel_6740\2910687795.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data[d] = pd.to_datetime(data[d])


In [16]:
data[date_cols].isnull().sum()

po_sent_to_vendor_date          5729
scheduled_delivery_date            0
delivered_to_client_date           0
delivery_recorded_date             0
pq_first_sent_to_client_date    5729
dtype: int64

In [17]:
# after converting to datetime data type there are now 5 cols with null values 
data.isnull().sum()

project_code                       0
pq_num                             0
po___so_num                        0
asn_dn_num                         0
country                            0
managed_by                         0
fulfill_via                        0
vendor_inco_term                   0
shipment_mode                    360
pq_first_sent_to_client_date    5729
po_sent_to_vendor_date          5729
scheduled_delivery_date            0
delivered_to_client_date           0
delivery_recorded_date             0
product_group                      0
sub_classification                 0
vendor                             0
item_description                   0
molecule_test_type                 0
brand                              0
dosage                          1734
dosage_form                        0
unit_of_measure_per_pack           0
line_item_quantity                 0
line_item_value                    0
pack_price                         0
unit_price                         0
m

In [18]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10320 entries, 1 to 86823
Data columns (total 32 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   project_code                  10320 non-null  object        
 1   pq_num                        10320 non-null  object        
 2   po___so_num                   10320 non-null  object        
 3   asn_dn_num                    10320 non-null  object        
 4   country                       10320 non-null  object        
 5   managed_by                    10320 non-null  object        
 6   fulfill_via                   10320 non-null  object        
 7   vendor_inco_term              10320 non-null  object        
 8   shipment_mode                 9960 non-null   object        
 9   pq_first_sent_to_client_date  4591 non-null   datetime64[ns]
 10  po_sent_to_vendor_date        4591 non-null   datetime64[ns]
 11  scheduled_delivery_date       103

In [19]:
# Stat Summary for 11 Numeric cols 
# 6 int/float datatypes and 5 datetime datatypes
data.describe().T

count                           mean  \
pq_first_sent_to_client_date     4591  2012-01-25 04:25:58.876061696   
po_sent_to_vendor_date           4591  2012-01-25 04:25:58.876061696   
scheduled_delivery_date         10320  2011-10-04 22:44:05.581395456   
delivered_to_client_date        10320  2011-09-28 22:06:50.232558080   
delivery_recorded_date          10320  2011-10-01 19:45:20.930232576   
unit_of_measure_per_pack      10320.0                      78.003682   
line_item_quantity            10320.0                   18339.276647   
line_item_value               10320.0                  157707.047954   
pack_price                    10320.0                      21.906249   
unit_price                    10320.0                       0.611634   
line_item_insurance_usd       10033.0                     240.205776   

                                              min                  25%  \
pq_first_sent_to_client_date  2006-04-19 00:00:00  2010-02-04 00:00:00   
po_sent_to_vendor_date        2006-04-19 00:00:00  2010-02-04 00:00:00   
scheduled_delivery_date       2006-05-02 00:00:00  2009-09-08 00:00:00   
delivered_to_client_date      2006-05-02 00:00:00  2009-09-03 00:00:00   
delivery_recorded_date        2006-05-02 00:00:00  2009-09-03 00:00:00   
unit_of_measure_per_pack                      1.0                 30.0   
line_item_quantity                            1.0                408.0   
line_item_value                               0.0              4318.65   
pack_price                                    0.0                 4.12   
unit_price                                    0.0                 0.08   
line_item_insurance_usd                       0.0                 6.51   

                                              50%                  75%  \
pq_first_sent_to_client_date  2012-06-06 00:00:00  2014-04-30 00:00:00   
po_sent_to_vendor_date        2012-06-06 00:00:00  2014-04-30 00:00:00   
scheduled_delivery_date       2011-12-05 12:00:00  2013-12-23 00:00:00   
delivered_to_client_date      2011-11-25 00:00:00  2013-12-23 00:00:00   
delivery_recorded_date        2011-11-25 00:00:00  2014-01-13 00:00:00   
unit_of_measure_per_pack                     60.0                 90.0   
line_item_quantity                         3000.0              17050.5   
line_item_value                           30504.0             166500.0   
pack_price                                    9.3                23.59   
unit_price                                   0.16               0.4625   
line_item_insurance_usd                     47.11                252.4   

                                              max            std  
pq_first_sent_to_client_date  2015-08-24 00:00:00            NaN  
po_sent_to_vendor_date        2015-08-24 00:00:00            NaN  
scheduled_delivery_date       2015-12-31 00:00:00            NaN  
delivered_to_client_date      2015-09-14 00:00:00            NaN  
delivery_recorded_date        2015-09-14 00:00:00            NaN  
unit_of_measure_per_pack                   1000.0      76.589256  
line_item_quantity                       619999.0    40041.59266  
line_item_value                         5951990.4  345346.992388  
pack_price                                1345.64      45.615097  
unit_price                                 238.65       3.276427  
line_item_insurance_usd                   7708.44     500.270659

In [20]:
# Stat Summary for 20 Categorical cols 
data.describe(include=object).T

count unique                                 top   freq
project_code        10320    142                          116-ZA-T30    768
pq_num              10320   1237                      Pre-PQ Process   2678
po___so_num         10320   6233                         SCMS-199289     67
asn_dn_num          10320   7030                           ASN-19166     54
country             10320     43                        South Africa   1406
managed_by          10320      4                            PMO - US  10261
fulfill_via         10320      2                            From RDC   5401
vendor_inco_term    10320      8                      N/A - From RDC   5401
shipment_mode        9960      4                                 Air   6109
product_group       10320      5                                 ARV   8548
sub_classification  10320      6                               Adult   6593
vendor              10320     73                       SCMS from RDC   5401
item_description    10320    184   Efavirenz 600mg, tablets, 30 Tabs    753
molecule_test_type  10320     86                           Efavirenz   1123
brand               10320     48                             Generic   7283
dosage               8586     54                               300mg    990
dosage_form         10320     17                              Tablet   3530
manufacturing_site  10320     88           Aurobindo Unit III, India   3171
weight_kilograms    10320   4688          Weight Captured Separately   1507
freight_cost_usd    10320   6733  Freight Included in Commodity Cost   1442

In [21]:
# Stat Summary for just the one Boolean datatype col (True or False)
data.describe(include=bool).T

count unique   top  freq
first_line_designation  10320      2  True  7030

## Identify all cols of object data type and clean/preprocess and convert data types as appropriate

In [22]:
# 20 cols that are categorical/objects
categ_data = data.select_dtypes(include=['object'])
categ_data

project_code          pq_num po___so_num asn_dn_num        country  \
id                                                                         
1       100-CI-T01  Pre-PQ Process      SCMS-4      ASN-8  Côte d'Ivoire   
3       108-VN-T01  Pre-PQ Process     SCMS-13     ASN-85        Vietnam   
4       100-CI-T01  Pre-PQ Process     SCMS-20     ASN-14  Côte d'Ivoire   
15      108-VN-T01  Pre-PQ Process     SCMS-78     ASN-50        Vietnam   
16      108-VN-T01  Pre-PQ Process     SCMS-81     ASN-55        Vietnam   
...            ...             ...         ...        ...            ...   
86818   103-ZW-T30       FPQ-15197    SO-50020    DN-4307       Zimbabwe   
86819   104-CI-T30       FPQ-15259    SO-50102    DN-4313  Côte d'Ivoire   
86821   110-ZM-T30       FPQ-14784    SO-49600    DN-4316         Zambia   
86822   200-ZW-T30       FPQ-16523    SO-51680    DN-4334       Zimbabwe   
86823   103-ZW-T30       FPQ-15197    SO-50022    DN-4336       Zimbabwe   

      managed_by  fulfill_via vendor_inco_term shipment_mode product_group  \
id                                                                           
1       PMO - US  Direct Drop              EXW           Air          HRDT   
3       PMO - US  Direct Drop              EXW           Air           ARV   
4       PMO - US  Direct Drop              FCA           Air          HRDT   
15      PMO - US  Direct Drop              EXW           Air           ARV   
16      PMO - US  Direct Drop              EXW           Air           ARV   
...          ...          ...              ...           ...           ...   
86818   PMO - US     From RDC   N/A - From RDC         Truck           ARV   
86819   PMO - US     From RDC   N/A - From RDC         Truck           ARV   
86821   PMO - US     From RDC   N/A - From RDC         Truck           ARV   
86822   PMO - US     From RDC   N/A - From RDC         Truck           ARV   
86823   PMO - US     From RDC   N/A - From RDC         Truck           ARV   

      sub_classification                                             vendor  \
id                                                                            
1               HIV test                        RANBAXY Fine Chemicals LTD.   
3              Pediatric                           Aurobindo Pharma Limited   
4               HIV test                               Abbott GmbH & Co. KG   
15                 Adult  SUN PHARMACEUTICAL INDUSTRIES LTD (RANBAXY LAB...   
16                 Adult                           Aurobindo Pharma Limited   
...                  ...                                                ...   
86818          Pediatric                                      SCMS from RDC   
86819              Adult                                      SCMS from RDC   
86821              Adult                                      SCMS from RDC   
86822              Adult                                      SCMS from RDC   
86823          Pediatric                                      SCMS from RDC   

                                        item_description  \
id                                                         
1      HIV, Reveal G3 Rapid HIV-1 Antibody Test, 30 T...   
3      Nevirapine 10mg/ml, oral suspension, Bottle, 2...   
4         HIV 1/2, Determine Complete HIV Kit, 100 Tests   
15                    Lamivudine 150mg, tablets, 60 Tabs   
16                     Stavudine 30mg, capsules, 60 Caps   
...                                                  ...   
86818  Lamivudine/Nevirapine/Zidovudine 30/50/60mg, d...   
86819  Lamivudine/Zidovudine 150/300mg, tablets, 60 Tabs   
86821  Efavirenz/Lamivudine/Tenofovir Disoproxil Fuma...   
86822  Lamivudine/Zidovudine 150/300mg, tablets, 60 Tabs   
86823  Lamivudine/Zidovudine 30/60mg, dispersible tab...   

                                      molecule_test_type      brand  \
id                                                                    
1               HIV, Reveal G3 Rapid HIV-1 Antibody

In [23]:
categ_data.columns

Index(['project_code', 'pq_num', 'po___so_num', 'asn_dn_num', 'country',
       'managed_by', 'fulfill_via', 'vendor_inco_term', 'shipment_mode',
       'product_group', 'sub_classification', 'vendor', 'item_description',
       'molecule_test_type', 'brand', 'dosage', 'dosage_form',
       'manufacturing_site', 'weight_kilograms', 'freight_cost_usd'],
      dtype='object')

# Categorical Data

In [24]:
# number of unique counts in categorical columns of object data type
print('The number of unique counts in each column:')
print()
for cols in categ_data.columns[0:]:
    print(cols,"=", len(categ_data[cols].unique()))

The number of unique counts in each column:

project_code = 142
pq_num = 1237
po___so_num = 6233
asn_dn_num = 7030
country = 43
managed_by = 4
fulfill_via = 2
vendor_inco_term = 8
shipment_mode = 5
product_group = 5
sub_classification = 6
vendor = 73
item_description = 184
molecule_test_type = 86
brand = 48
dosage = 55
dosage_form = 17
manufacturing_site = 88
weight_kilograms = 4688
freight_cost_usd = 6733


In [25]:
categ_data.isnull().sum() 

project_code             0
pq_num                   0
po___so_num              0
asn_dn_num               0
country                  0
managed_by               0
fulfill_via              0
vendor_inco_term         0
shipment_mode          360
product_group            0
sub_classification       0
vendor                   0
item_description         0
molecule_test_type       0
brand                    0
dosage                1734
dosage_form              0
manufacturing_site       0
weight_kilograms         0
freight_cost_usd         0
dtype: int64

In [26]:
# define function to split cols with identification numbers
def ident(data):
    data_split = data.split("-")
    data_return = data_split[0]
    return data_return

In [27]:
data['pq_num'].apply(ident).value_counts()

pq_num
FPQ    7642
Pre    2678
Name: count, dtype: int64

In [28]:
data['po___so_num'].apply(ident).value_counts()

po___so_num
SO      5401
SCMS    4918
DSCM       1
Name: count, dtype: int64

Only two types: FPQ and Pre-PQ

In [29]:
data['asn_dn_num'].apply(ident).value_counts()

asn_dn_num
DN     5401
ASN    4919
Name: count, dtype: int64

5401 products were shipped from Direct Drop deliveries based on the DN number. 4919 deliveries were made from Regional Distribution Centers so there was a middleman between the supplier and buyer during transporation, based on the ASN number.

## project_code, pq_num, po___so_num, asn_dn_num

1) project_code = Project Code. A specific number for a project. If there are multiple of the same project codes, they were delivered together to the same destination.

2) pq_num = Price quote (PQ) number.

3) po___so_num = Purchase Order/Sales Order number. 

4) asn_dn_num = Shipment number: Advanced Shipment Note (ASN) for Direct Drop deliveries, or Delivery Note (DN) for from Regional Distribution Centers (RDC) deliveries. RDCs are large warehouse facilities that streamline the process of getting the products to the buyers faster.

These are basically identification numbers to keep track of sales and shipment. There are no null or missing values in these columns. While there are duplicates, it makes sense for there to be multiple products in one order, so Advanced Shipment Note (ASN) and Delivery Note (DN) numbers would be greater.

In [30]:
data['project_code'].unique() # 142 unique product code numbers

array(['100-CI-T01', '108-VN-T01', '112-NG-T01', '110-ZM-T01',
       '109-TZ-T01', '102-NG-T01', '107-RW-T01', '106-HT-T01',
       '113-ZW-T01', '108-VN-T30', '104-CI-T01', '100-HT-T01',
       '117-ET-T01', '116-ZA-T01', '123-NG-T01', '125-HT-T01',
       '102-GY-T01', '119-NA-T01', '131-NG-T01', '102-BW-T01',
       '111-MZ-T01', '144-BW-T01', '102-KE-T01', '133-NG-T01',
       '100-KZ-T01', '141-NA-T01', '114-UG-T01', '105-GY-T01',
       '139-NA-T01', '129-KG-T01', '100-SN-T01', '128-BJ-T01',
       '102-LS-T01', '130-NG-T01', '100-BW-T01', '100-ZW-T01',
       '100-PK-T01', '126-NG-T01', '151-NG-T01', '100-SZ-T01',
       '100-GH-T01', '120-AO-T01', '132-NG-T01', '153-NG-T01',
       '100-LB-T01', '151-NG-T30', '127-KE-T01', '510-KE-T01',
       '100-SL-T01', '136-RW-T01', '102-KE-T30', '110-ZM-T30',
       '106-HT-T30', '105-SS-T30', '111-MZ-T30', '102-BI-T30',
       '122-HT-T30', '161-ZA-T30', '116-ZA-T30', '133-NG-T30',
       '103-DO-T30', '104-CI-T30', '107-RW-T30', '103-M

In [31]:
data.loc[data['project_code'] == '153-NA-T30']

project_code    pq_num  po___so_num asn_dn_num  country managed_by  \
id                                                                         
22726   153-NA-T30  FPQ-8072  SCMS-117210  ASN-12115  Namibia   PMO - US   
23023   153-NA-T30  FPQ-9698  SCMS-144310  ASN-15069  Namibia   PMO - US   
34698   153-NA-T30  FPQ-8795  SCMS-131120  ASN-14198  Namibia   PMO - US   
39556   153-NA-T30  FPQ-8072  SCMS-117200  ASN-12163  Namibia   PMO - US   
71391   153-NA-T30  FPQ-9698  SCMS-144300  ASN-15068  Namibia   PMO - US   
75091   153-NA-T30  FPQ-8795  SCMS-131100  ASN-13763  Namibia   PMO - US   
80299   153-NA-T30  FPQ-8795  SCMS-131110  ASN-14106  Namibia   PMO - US   

       fulfill_via vendor_inco_term shipment_mode  \
id                                                  
22726  Direct Drop              EXW           Air   
23023  Direct Drop              EXW           Air   
34698  Direct Drop              EXW           Air   
39556  Direct Drop              EXW           Air   
71391  Direct Drop              EXW           Air   
75091  Direct Drop              EXW           Air   
80299  Direct Drop              EXW           Air   

      pq_first_sent_to_client_date  ... unit_of_measure_per_pack  \
id                                  ...                            
22726                   2011-08-03  ...                      100   
23023                   2012-05-02  ...                      100   
34698                   2012-01-12  ...                       25   
39556                   2011-08-03  ...                       20   
71391                   2012-05-02  ...                       20   
75091                   2012-01-12  ...                       20   
80299                   2012-01-12  ...                      100   

      line_item_quantity line_item_value pack_price unit_price  \
id                                                               
22726                150         11655.0       77.7       0.78   
23023                 75          5827.5       77.7       0.78   
34698                 50          4750.0       95.0       3.80   
39556                750         24000.0       32.0       1.60   
71391                416         13312.0       32.0       1.60   
75091                728         23296.0       32.0       1.60   
80299                150         11655.0       77.7       0.78   

                  manufacturing_site first_line_designation weight_kilograms  \
id                                                                             
22726                Inverness Japan                   True               36   
23023                Inverness Japan                   True               19   
34698  Chembio Diagnostics Sys. Inc.                   True               36   
39556           Trinity Biotech, Plc                   True              248   
71391           Trinity Biotech, Plc                   True              249   
75091           Trinity Biotech, Plc                   True              238   
80299                Inverness Japan                   True               35   

      freight_cost_usd line_item_insurance_usd  
id                                              
22726           859.18                   19.23  
23023           718.32                    8.17  
34698           481.63                    6.66  
39556          6654.24                   39.60  
71391          1640.99                   18.66  
75091          2789.67                   32.66  
80299           840.46                   16.34  

[7 rows x 32 columns]

Project code 153-NA-T30 contain commodities that were sent together to the same destination, the country of Namibia. However, they were delivered at different times within the time period of 9/20/2011-07/16/2012 (around 10 months). This could be explained by the vast country of Namibia

Project codes with the same pq_sum were delivered together to the same facility or location. For example, pq_num of FPQ-8072 had 2 orders with a Price Quote Number of SCMS-117210 & SCMS-117200. The Purchase Order/Sales Order numbers are ASN-12115 and ASN-12163, respectively. 

1. SCMS-117210 - ASN-12115 - 9/20/2011

2. SCMS-117200 - ASN-12163 - 10/5/2011

While both orders had their Price Quote Number sent from the buyer to the vendor on 8/3/2011, the actual products were delivered on 9/20/2011 and 10/5/2011, respectively. This makes sense because they could have been delivered to different health facilities that were close by each other, but many factors such as weather, wildlife, etc. could impact transporation.

## country col
5) country = list of supported countries that receive health commodities (Antiretroviral (ARV) and HIV lab shipments) to treat HIV/AIDS

In [32]:
data['country'].unique()

array(["Côte d'Ivoire", 'Vietnam', 'Nigeria', 'Zambia', 'Tanzania',
       'Rwanda', 'Haiti', 'Zimbabwe', 'Ethiopia', 'South Africa',
       'Guyana', 'Namibia', 'Botswana', 'Mozambique', 'Kenya',
       'Kazakhstan', 'Uganda', 'Kyrgyzstan', 'Senegal', 'Benin',
       'Lesotho', 'Pakistan', 'Swaziland', 'Ghana', 'Angola', 'Lebanon',
       'Sierra Leone', 'Cameroon', 'South Sudan', 'Burundi',
       'Dominican Republic', 'Malawi', 'Congo, DRC', 'Sudan', 'Mali',
       'Guatemala', 'Togo', 'Afghanistan', 'Liberia', 'Burkina Faso',
       'Guinea', 'Libya', 'Belize'], dtype=object)

This lists all countries in dataset and we can see the dataset primarily consists of shipments to the African continent and some other countries in Asia. We see there are 43 unique countries in the dataset and no duplicates or typos are present.

In [33]:
# get top 5 countries
data['country'].value_counts().nlargest(5)

country
South Africa     1406
Nigeria          1194
Côte d'Ivoire    1083
Uganda            779
Vietnam           688
Name: count, dtype: int64

After getting the number of rows/counts for each country, we see most of the shipments go to South Africa, Nigeria, Côte d'Ivoire. Each country has over 1000 products shipped to them.

According to https://cgdv.github.io/achive/viz_111/USAID_Supply_Chain/index.html, SCMS (Supply Chain Management System) is the largest supplier/distributor of ARVs (Anti-Retroviral Treatment) to treat HIV ever since 2005. They have distributed over 1 billion dollars worth of ARVs to these supported countries. 

Ever since 2006, USAID has delivered over 350 million dollars of ARV and HIV lab supplies to Nigeria, which is one of the largest countries in Africa. South Africa has the most number of people in the world that have HIV/AIDS, while Nigeria has the second highest number of individuals that live with HIV/AIDS. 

On the other hand, over 246 million dollars worth of supplies were sent to Zambia. Even though Zambia is not one of the top 5 countries with the most shipments, according to the literature Zambia is known to be undersupported, which makes sense with this data.

## managed_by, fulfill_via, vendor_inco_term
6) managed_by = SCMS managing office: either the Program Management Office (PMO) in the U.S. or the relevant SCMS field office. Either the order was managed by one of the 4 offices: PMO - US, South Africa Field Office, Haiti Field Office, or the Ethiopia Field Office.

7) fulfill_via = Method through which the shipment was fulfilled: either one of the two options, Direct Drop from vendor or from stock available in the Regional Distribution Centers (RDCs). If shipment was a direct drop, this means the seller has their suppliers or wholesalers to directly deliver orders to customers on their behalf. According to https://geo2.com/what-is-rdc-delivery-regional-distribution-centres/, RDCs are part of the distribution process and temporarily store commodities before being repacked and shipped later. This saves transpotation costs. So, when the product was fulfilled via RDCs, this means it came from the central facility warehouse, then to the RDC, and finally delivered to the customer.

8) vendor_inco_term = The vendor INCO term (also known as International Commercial Terms) for Direct Drop deliveries. According to https://www.trade.gov/know-your-incoterms#:~:text=Incoterms%2C%20widely%2Dused%20terms%20of,clearance%2C%20and%20other%20logistical%20activities. Incoterms clarify the terms, rules, and responsibilities of both parties in a sales transaction. There are 7 Incoterms for any shipped mode of transport. 

    1) EXW - Ex Works (insert place of delivery) 
    2) FCA  - Free Carrier (Insert named place of delivery) 
    3) CPT  - Carriage Paid to (insert place of destination) 
    4) CIP -  Carriage and Insurance Paid To (insert place of destination) 
    5) DAP - Delivered at Place (insert named place of destination) 
    6) DPU - Delivered at Place Unloaded (insert of place of destination) 
    7) DDP - Delivered Duty Paid (Insert place of destination).


In [34]:
data['managed_by'].value_counts()

managed_by
PMO - US                     10261
South Africa Field Office       57
Haiti Field Office               1
Ethiopia Field Office            1
Name: count, dtype: int64

In [35]:
data['fulfill_via'].value_counts()

fulfill_via
From RDC       5401
Direct Drop    4919
Name: count, dtype: int64

In [36]:
data['vendor_inco_term'].value_counts()

vendor_inco_term
N/A - From RDC    5401
EXW               2777
DDP               1443
FCA                397
CIP                275
DDU                 15
DAP                  9
CIF                  3
Name: count, dtype: int64

There are 7 global Incoterms used for any mode of transport. For this dataset there is an 8th value 'N/A - From RDC'. This means there is no Incoterm in the transaction because it was from a RDC. We see 5404 transactions came from a RDC.

In [37]:
data.loc[data['vendor_inco_term'] == 'N/A - From RDC']

project_code          pq_num po___so_num asn_dn_num        country  \
id                                                                         
10809   107-RW-T01  Pre-PQ Process      SO-249       DN-6         Rwanda   
11100   110-ZM-T01  Pre-PQ Process      SO-310     DN-122         Zambia   
10534   110-ZM-T01  Pre-PQ Process      SO-220       DN-4         Zambia   
10535   111-MZ-T01  Pre-PQ Process      SO-321      DN-13     Mozambique   
10536   111-MZ-T01  Pre-PQ Process      SO-321      DN-13     Mozambique   
...            ...             ...         ...        ...            ...   
86818   103-ZW-T30       FPQ-15197    SO-50020    DN-4307       Zimbabwe   
86819   104-CI-T30       FPQ-15259    SO-50102    DN-4313  Côte d'Ivoire   
86821   110-ZM-T30       FPQ-14784    SO-49600    DN-4316         Zambia   
86822   200-ZW-T30       FPQ-16523    SO-51680    DN-4334       Zimbabwe   
86823   103-ZW-T30       FPQ-15197    SO-50022    DN-4336       Zimbabwe   

      managed_by fulfill_via vendor_inco_term shipment_mode  \
id                                                            
10809   PMO - US    From RDC   N/A - From RDC           Air   
11100   PMO - US    From RDC   N/A - From RDC           NaN   
10534   PMO - US    From RDC   N/A - From RDC           NaN   
10535   PMO - US    From RDC   N/A - From RDC           Air   
10536   PMO - US    From RDC   N/A - From RDC           Air   
...          ...         ...              ...           ...   
86818   PMO - US    From RDC   N/A - From RDC         Truck   
86819   PMO - US    From RDC   N/A - From RDC         Truck   
86821   PMO - US    From RDC   N/A - From RDC         Truck   
86822   PMO - US    From RDC   N/A - From RDC         Truck   
86823   PMO - US    From RDC   N/A - From RDC         Truck   

      pq_first_sent_to_client_date  ... unit_of_measure_per_pack  \
id                                  ...                            
10809                          NaT  ...                       60   
11100                          NaT  ...                       60   
10534                          NaT  ...                      270   
10535                          NaT  ...                      270   
10536                          NaT  ...                       30   
...                            ...  ...                      ...   
86818                          NaT  ...                       60   
86819                          NaT  ...                       60   
86821                          NaT  ...                       30   
86822                          NaT  ...                       60   
86823                          NaT  ...                       60   

      line_item_quantity line_item_value pack_price unit_price  \
id                                                               
10809                600        21480.00      35.80       0.60   
11100               1000        35800.00      35.80       0.60   
10534                228        17948.16      78.72       0.29   
10535                764        60814.40      79.60       0.29   
10536                341         1166.22       3.42       0.11   
...                  ...             ...        ...        ...   
86818             166571       599655.60       3.60       0.06   
86819              21072       137389.44       6.52       0.11   
86821             514526      5140114.74       9.99       0.33   
86822              17465       113871.80       6.52       0.11   
86823              36639        72911.61       1.99       0.03   

                   manufacturing_site first_line_designation  \
id                                                             
10809       Aurobindo Unit III, India                   True   
11100       Aurobindo Unit III, India                   True   
10534                    Roche Madrid                   True   
10535                    Roche Madrid                   True   
10536  MSD Manati, Puerto Rico, (USA)                  False   
...  

## shipment_mode
9) shipment_mode = Method by which commodities are shipped

In [38]:
data['shipment_mode'].value_counts()

shipment_mode
Air            6109
Truck          2830
Air Charter     650
Ocean           371
Name: count, dtype: int64

There are 4 unique modes of transportation: Air, Truck, Air Charter, or Ocean. However, there are 360 nan values

In [39]:
data[data['shipment_mode'].isnull()]

project_code          pq_num po___so_num asn_dn_num       country  \
id                                                                        
69      102-NG-T01  Pre-PQ Process    SCMS-354    ASN-608       Nigeria   
262     116-ZA-T01  Pre-PQ Process  SCMS-14050   ASN-1251  South Africa   
343     116-ZA-T01  Pre-PQ Process  SCMS-16600   ASN-1520  South Africa   
400     117-ET-T01  Pre-PQ Process  SCMS-18430   ASN-1465      Ethiopia   
459     116-ZA-T01  Pre-PQ Process  SCMS-21460   ASN-1667  South Africa   
...            ...             ...         ...        ...           ...   
12740   110-ZM-T01  Pre-PQ Process     SO-1078     DN-438        Zambia   
12761   114-UG-T01  Pre-PQ Process     SO-1820     DN-622        Uganda   
12774   114-UG-T01  Pre-PQ Process     SO-1082     DN-589        Uganda   
12789   117-ET-T01  Pre-PQ Process     SO-1950     DN-669      Ethiopia   
12835   109-TZ-T01  Pre-PQ Process     SO-1697     DN-893      Tanzania   

      managed_by  fulfill_via vendor_inco_term shipment_mode  \
id                                                             
69      PMO - US  Direct Drop              CIP           NaN   
262     PMO - US  Direct Drop              DDP           NaN   
343     PMO - US  Direct Drop              DDP           NaN   
400     PMO - US  Direct Drop              DDP           NaN   
459     PMO - US  Direct Drop              DDP           NaN   
...          ...          ...              ...           ...   
12740   PMO - US     From RDC   N/A - From RDC           NaN   
12761   PMO - US     From RDC   N/A - From RDC           NaN   
12774   PMO - US     From RDC   N/A - From RDC           NaN   
12789   PMO - US     From RDC   N/A - From RDC           NaN   
12835   PMO - US     From RDC   N/A - From RDC           NaN   

      pq_first_sent_to_client_date  ... unit_of_measure_per_pack  \
id                                  ...                            
69                             NaT  ...                      100   
262                     2007-11-12  ...                      200   
343                     2008-01-04  ...                      200   
400                     2008-02-29  ...                      100   
459                     2008-04-04  ...                       60   
...                            ...  ...                      ...   
12740                          NaT  ...                       60   
12761                          NaT  ...                       60   
12774                          NaT  ...                       60   
12789                          NaT  ...                       60   
12835                          NaT  ...                       60   

      line_item_quantity line_item_value pack_price unit_price  \
id                                                               
69                  1500       120000.00      80.00       0.80   
262                   40          317.60       7.94       0.04   
343                  200          340.00       1.70       0.01   
400                   63        13399.25     212.69       2.13   
459                    6           77.10      12.85       0.21   
...                  ...             ...        ...        ...   
12740              18974       140407.60       7.40       0.12   
12761               1860        13764.00       7.40       0.12   
12774                328         1115.20       3.40       0.06   
12789              78460       119259.20       1.52       0.03   
12835               5730       157746.90      27.53       0.46   

                  manufacturing_site first_line_designation  \
id                                                            
69     ABBVIE GmbH & Co.KG Wiesbaden                   True   
262         GSK Mississauga (Canada)                   True   
343               BMS Meymac, France                   True   
400             Trinity Biotech, Plc                   True   
459               BMS Meymac, France                   True   
...         

360 rows are missing the shipment mode, which could be explained by human error and the mode of transport was not recorded properly.

In [40]:
####print(data[data['shipment_mode'].isnull()] and )

The majority of shipment mode nan values correlate with the incoterm 'DDP' (Delivered Duty Paid) or 'N/A - From RDC'. 

DPP means when a good is bought or sold the seller delivers the goods to a location that both parties agreed upon. The location of transaction can be any where, according to https://www.customssupport.com/insights/incoterms-explained-delivery-duty-paid-ddp#:~:text=When%20goods%20are%20bought%20or,needs%20to%20be%20specifically%20named. Therefore, it makes sense that the shipment_mode was left blank because the mode of transport wouldn't fall under 'Air', 'Truck', 'Air Charter', or 'Ocean'. Similarly for when the commodity was fulfilled from the RDC. Then there is no need for an incoterm. Perhaps, another mode of transportation was used from the RDC to the buyers location.

## product_group, sub_classification

10) product_group = Product group for item, i.e. ARV, HRDT. There are 5 product groups the commodities belong to. According to https://cgdv.github.io/achive/viz_111/USAID_Supply_Chain/index.html:

    1. HRDT = HIV Rapid Diagnostic Test (HIV)
    2. ARV = Anti-Retroviral Treatment = prevent HIV infection from replication (HIV)
    3. ACT = Artemisinin-based Combination Therapy (Malaria)
    4. MRDT = Malarial Rapid Diagnostic Test (Malaria)
    5. ANTM = Anti-malarial medicine (Malaria)

11) sub_classification = Identifies relevant product sub classifications, such as whether ARVs are pediatric or adult, whether a malaria product is an artemisinin-based combination therapy (ACT), etc.

    1. Adult = ARVs for adults (HIV)
    2. Pediatric = ARVs for children/under 18 (HIV)
    3. HIV test = given a blood/fluid sample, detects antibodies/antigens to diagnose HIV 
    4. HIV test - Ancillary = an additional test to diagnose HIV test, utilizing a special technique of immunofluorescence staining
    5. Malaria = malaria product (a rapid test or anti-malarial medicine)
    6. ACT = Artemisinin-based Combination Therapy (Malaria)

In [41]:
data['product_group'].value_counts()

product_group
ARV     8548
HRDT    1726
ANTM      22
ACT       16
MRDT       8
Name: count, dtype: int64

In [42]:
data['sub_classification'].value_counts()

sub_classification
Adult                   6593
Pediatric               1955
HIV test                1565
HIV test - Ancillary     161
Malaria                   30
ACT                       16
Name: count, dtype: int64

There are no null or missing values in both product_group and sub_classification columns. 10274 products in this dataset are used to treat or diagnose HIV (basically 99.55%), and 46 for malaria (0.0044%) accounts for the entire dataset.

## vendor, item_description, molecule_test_type, brand, dosage, dosage_form, manufacturing_site, weight_kilograms, freight_cost_usd

12) vendor = name of the vendor/supplier that sells the product

13) item_description = Product name and formulation from Partnership for Supply Chain Management (PFSCM) Item Master

14) molecule_test_type = whether the product are active drug(s) or a type of test kit

15) brand = Generic or branded name for the item

16) dosage = Item dosage and unit

17) dosage_form = Dosage form for the item (tablet, oral solution, injection, etc.). There are 17 forms that the product can be taken

    1. Tablet = a compressed pill in solid form entirely made of medication. There is no outer coating
    2. Tablet - FDC = Fixed-Dose Combination (FDC) tablet. Two or more drugs contained in a single dosage form. This aims to reduce the amount of pills one needs to take everyday. For example, one HIV drug is Atripla (a combination of efavirenz, emtricitabine, and tenofovir disoproxil fumarate) according to https://clinicalinfo.hiv.gov/en/glossary/fixed-dose-combination-fdc.
    3. Test kit = 
    4. Capsule = 
    5. Oral solution = 
    6. Chewable/dispersible tablet - FDC = 
    7. Oral suspension = 
    8. Test kit - Ancillary = 
    9. Chewable/dispersible tablet = 
    10. Delayed-release capsules
    11. Delayed-release capsules - blister
    12. Powder for oral solution
    13. Tablet - FDC + co-blister
    14. Tablet - FDC + blister
    15. Tablet - blister
    16. Injection 
    17. Oral powder

18) manufacturing_site = Identifies manufacturing site for the line item for direct drop and from RDC deliveries

19) weight_kilograms = Weight for all lines on an ASN DN

20) freight_cost_usd = Freight charges associated with all lines on the respective ASN DN


In [43]:
data['dosage_form'].value_counts()

dosage_form
Tablet                                3530
Tablet - FDC                          2749
Test kit                              1573
Capsule                                729
Oral solution                          727
Chewable/dispersible tablet - FDC      239
Oral suspension                        214
Test kit - Ancillary                   161
Chewable/dispersible tablet            146
Delayed-release capsules               131
Delayed-release capsules - blister      41
Powder for oral solution                28
Tablet - FDC + co-blister               20
Tablet - FDC + blister                  15
Tablet - blister                        10
Injection                                6
Oral powder                              1
Name: count, dtype: int64

# Clean categorical attributes with noisy text fields

Clean these 5 categorical attributes with noisy text fields: ['item_description', 'molecule_test_type', 'manufacturing_site', 'weight_kilograms', 'freight_cost_usd']

In [44]:
data[['item_description', 'molecule_test_type', 'manufacturing_site', 'weight_kilograms', 'freight_cost_usd']]

item_description  \
id                                                         
1      HIV, Reveal G3 Rapid HIV-1 Antibody Test, 30 T...   
3      Nevirapine 10mg/ml, oral suspension, Bottle, 2...   
4         HIV 1/2, Determine Complete HIV Kit, 100 Tests   
15                    Lamivudine 150mg, tablets, 60 Tabs   
16                     Stavudine 30mg, capsules, 60 Caps   
...                                                  ...   
86818  Lamivudine/Nevirapine/Zidovudine 30/50/60mg, d...   
86819  Lamivudine/Zidovudine 150/300mg, tablets, 60 Tabs   
86821  Efavirenz/Lamivudine/Tenofovir Disoproxil Fuma...   
86822  Lamivudine/Zidovudine 150/300mg, tablets, 60 Tabs   
86823  Lamivudine/Zidovudine 30/60mg, dispersible tab...   

                                      molecule_test_type  \
id                                                         
1               HIV, Reveal G3 Rapid HIV-1 Antibody Test   
3                                             Nevirapine   
4                    HIV 1/2, Determine Complete HIV Kit   
15                                            Lamivudine   
16                                             Stavudine   
...                                                  ...   
86818                   Lamivudine/Nevirapine/Zidovudine   
86819                              Lamivudine/Zidovudine   
86821  Efavirenz/Lamivudine/Tenofovir Disoproxil Fuma...   
86822                              Lamivudine/Zidovudine   
86823                              Lamivudine/Zidovudine   

                   manufacturing_site            weight_kilograms  \
id                                                                  
1          Ranbaxy Fine Chemicals LTD                          13   
3           Aurobindo Unit III, India                         358   
4       ABBVIE GmbH & Co.KG Wiesbaden                         171   
15      Ranbaxy, Paonta Shahib, India                        1855   
16          Aurobindo Unit III, India                        7590   
...                               ...                         ...   
86818      Mylan,  H-12 & H-13, India     See DN-4307 (ID#:83920)   
86819    Hetero Unit III Hyderabad IN     See DN-4313 (ID#:83921)   
86821   Cipla Ltd A-42 MIDC Mahar. IN  Weight Captured Separately   
86822  Mylan (formerly Matrix) Nashik                        1392   
86823               Cipla, Goa, India  Weight Captured Separately   

                         freight_cost_usd  
id                                         
1                                  780.34  
3                                  4521.5  
4                                 1653.78  
15                               16007.06  
16                               45450.08  
...                                   ...  
86818             See DN-4307 (ID#:83920)  
86819             See DN-4313 (ID#:83921)  
86821  Freight Included in Commodity Cost  
86822  Freight Included in Commodity Cost  
86823  Freight Included in Commodity Cost  

[10320 rows x 5 columns]

# 1. Item Description

In [45]:
# sort alphabetically to view
# item_description = 184 unique items 
itemd = data.sort_values('item_description')
itemd['item_description'].unique()

array(['#102198**Didanosine 200mg [Videx], tablets, 60 Tabs',
       '#108853** HIV, Genie II HIV-1/HIV-2 Kit, 40 Tests',
       '#109435**HIV, OraQuick Controls Kit, 3 Vials',
       '#109436**HIV 1, Uni-Gold Recombigen HIV Control, Vial 2 x 0.5 ml',
       '#109440**HIV 1/2, INSTI HIV Antibody Individual POC, 1 Test',
       'Abacavir 20mg/ml [Ziagen], oral solution, Bottle, 240 ml',
       'Abacavir 20mg/ml, [DON] oral solution, Bottle, 240 ml',
       'Abacavir 20mg/ml, oral solution w/syringe, Bottle, 240 ml',
       'Abacavir 20mg/ml, oral solution, Bottle, 240 ml',
       'Abacavir 300mg [Ziagen], tablets, Blister 6 x 10 Tabs',
       'Abacavir 300mg, tablets, 60 Tabs',
       'Abacavir 60mg, dispersible tablets, 60 Tabs',
       'Abacavir/Lamivudine 60/30mg, dispersible tablets, 60 Tabs',
       'Abacavir/Lamivudine 60/30mg, tablets, 30 Tabs',
       'Abacavir/Lamivudine 60/30mg, tablets, 60 Tabs',
       'Abacavir/Lamivudine 600/300mg, scored tablets, 30 Tabs',
       'Abacavi

'item_description' & 'molecule_test_type' columns are almost identical. The item description is more detailed containing information about the dosage and unit of measure (per pack). These fields are already present as other columns in the dataset: 'molecule_test_type', 'dosage', 'dosage_form', & 'unit of measure (per pack)' That's why there are missing fields because some of the item descriptions do not contain the dosage and only the dosage form.

item_description = molecule_test_type, brand, dosage, dosage_form, unit_of_measure_per_pack

This, this col can be dropped

In [46]:
data['item_description'][1]

'HIV, Reveal G3 Rapid HIV-1 Antibody Test, 30 Tests'

In [47]:
data.iloc[[0], [17, 18, 19, 20, 21, 22]]

item_description  \
id                                                      
1   HIV, Reveal G3 Rapid HIV-1 Antibody Test, 30 T...   

                          molecule_test_type   brand dosage dosage_form  \
id                                                                        
1   HIV, Reveal G3 Rapid HIV-1 Antibody Test  Reveal    NaN    Test kit   

    unit_of_measure_per_pack  
id                            
1                         30

'''# split item_description col to 3 cols
1st col is the brand/drug name, the active ingredients contained in the medication
 2nd col is the brief description of the item
 3rd col is the amount in each item
#data[['active_ingredients', 'drug_description', 'amounts']] = data['item_description'].str.split(',', expand=True)'''

In [48]:
# drop item_description column 
data.drop("item_description",axis=1,inplace=True)
data

C:\Users\Bertha\AppData\Local\Temp\ipykernel_6740\3533548080.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data.drop("item_description",axis=1,inplace=True)


project_code          pq_num po___so_num asn_dn_num        country  \
id                                                                         
1       100-CI-T01  Pre-PQ Process      SCMS-4      ASN-8  Côte d'Ivoire   
3       108-VN-T01  Pre-PQ Process     SCMS-13     ASN-85        Vietnam   
4       100-CI-T01  Pre-PQ Process     SCMS-20     ASN-14  Côte d'Ivoire   
15      108-VN-T01  Pre-PQ Process     SCMS-78     ASN-50        Vietnam   
16      108-VN-T01  Pre-PQ Process     SCMS-81     ASN-55        Vietnam   
...            ...             ...         ...        ...            ...   
86818   103-ZW-T30       FPQ-15197    SO-50020    DN-4307       Zimbabwe   
86819   104-CI-T30       FPQ-15259    SO-50102    DN-4313  Côte d'Ivoire   
86821   110-ZM-T30       FPQ-14784    SO-49600    DN-4316         Zambia   
86822   200-ZW-T30       FPQ-16523    SO-51680    DN-4334       Zimbabwe   
86823   103-ZW-T30       FPQ-15197    SO-50022    DN-4336       Zimbabwe   

      managed_by  fulfill_via vendor_inco_term shipment_mode  \
id                                                             
1       PMO - US  Direct Drop              EXW           Air   
3       PMO - US  Direct Drop              EXW           Air   
4       PMO - US  Direct Drop              FCA           Air   
15      PMO - US  Direct Drop              EXW           Air   
16      PMO - US  Direct Drop              EXW           Air   
...          ...          ...              ...           ...   
86818   PMO - US     From RDC   N/A - From RDC         Truck   
86819   PMO - US     From RDC   N/A - From RDC         Truck   
86821   PMO - US     From RDC   N/A - From RDC         Truck   
86822   PMO - US     From RDC   N/A - From RDC         Truck   
86823   PMO - US     From RDC   N/A - From RDC         Truck   

      pq_first_sent_to_client_date  ... unit_of_measure_per_pack  \
id                                  ...                            
1                              NaT  ...                       30   
3                              NaT  ...                      240   
4                              NaT  ...                      100   
15                             NaT  ...                       60   
16                             NaT  ...                       60   
...                            ...  ...                      ...   
86818                          NaT  ...                       60   
86819                          NaT  ...                       60   
86821                          NaT  ...                       30   
86822                          NaT  ...                       60   
86823                          NaT  ...                       60   

      line_item_quantity line_item_value pack_price unit_price  \
id                                                               
1                     19          551.00      29.00       0.97   
3                   1000         6200.00       6.20       0.03   
4                    500        40000.00      80.00       0.80   
15                 31920       127360.80       3.99       0.07   
16                 38000       121600.00       3.20       0.05   
...                  ...             ...        ...        ...   
86818             166571       599655.60       3.60       0.06   
86819              21072       137389.44       6.52       0.11   
86821             514526      5140114.74       9.99       0.33   
86822              17465       113871.80       6.52       0.11   
86823              36639        72911.61       1.99       0.03   

                   manufacturing_site first_line_designation  \
id                                                             
1          Ranbaxy Fine Chemicals LTD                   True   
3           Aurobindo Unit III, India                   True   
4       ABBVIE GmbH & Co.KG Wiesbaden                   True   
15      Ranbaxy, Paonta Shahib, India                   True   
16          Aurobindo Unit III, India                   

https://stackoverflow.com/questions/73091240/python-pandas-search-string-in-a-column-and-get-initial-position-if-the-string-i

In [49]:
# oral solution w/syringe, Bottle
#data.loc[data['item_description'] == 'bottle']
#data['item_description'].str.contains('Bottle').any()

item description column is more descriptive, contains relevant information on whether the product comes in a bottle, vial, if it comes with a syringe, etc.

Examples:
oral solution w/syringe, Bottle
oral solution, Bottle
solution for IV infusion injection, Vial 
oral solution, cool, Bottle,
firm coated tablets


In [50]:
'''search_bottle = "bottle"
data['bottle'] = data['item_description'].apply(lambda x: x.lower().index(search_bottle.lower()) +1 if search_bottle.lower() in x.lower() else "NA")
print(data)'''

'search_bottle = "bottle"\ndata[\'bottle\'] = data[\'item_description\'].apply(lambda x: x.lower().index(search_bottle.lower()) +1 if search_bottle.lower() in x.lower() else "NA")\nprint(data)'

## 2. Molecule Test Type

86 unique test types

In [51]:
data['molecule_test_type'].value_counts()

molecule_test_type
Efavirenz                           1123
Nevirapine                           877
Lamivudine/Nevirapine/Zidovudine     707
Lamivudine/Zidovudine                689
Lopinavir/Ritonavir                  633
                                    ... 
Mefloquine (as hydrochloride)          1
Quinine (as sulfate)                   1
Primaquine base (as diphosphate)       1
Lamivudine/Zidovudine+Abacavir         1
Nelfinavir                             1
Name: count, Length: 86, dtype: int64

In [52]:
# sort alphabetically to view
mtt = data.sort_values('molecule_test_type')
mtt['molecule_test_type'].unique()

array(['Abacavir', 'Abacavir/Lamivudine',
       'Abacavir/Lamivudine/Zidovudine', 'Artemether/Lumefantrine',
       'Artesunate+Amodiaquine', 'Atazanavir', 'Atazanavir/Ritonavir',
       'Chase Buffer, Determine, 100 Tests, 2.5ml x 1 Vial',
       'Chloroquine base (as phosphate)', 'Darunavir', 'Didanosine',
       'Didanosine EC', 'Efavirenz',
       'Efavirenz/Emtricitabine/Tenofovir Disoproxil Fumarate',
       'Efavirenz/Lamivudine/Tenofovir Disoproxil Fumarate',
       'Emtricitabine/Tenofovir Disoproxil Fumarate', 'Etravirine',
       'HIV 1, Uni-Gold Recombigen HIV Control, Vial 2 x 0.5 ml',
       'HIV 1/2, Bioline 3.0 Diluent, 25 Tests',
       'HIV 1/2, Bioline 3.0 Kit',
       'HIV 1/2, Bioline 3.0 Kit (no accessories)',
       'HIV 1/2, Bioline 3.0 Kit, Lancets, Capillary pipets, Alcohol swabs included',
       'HIV 1/2, Bundi Rapid HIV Kit', 'HIV 1/2, Capillus HIV Kit',
       'HIV 1/2, Clearview Complete Kit',
       'HIV 1/2, Colloidal Gold, Diagnostic Kit Set (includes

In [53]:
data['active_ingredients'] = data['molecule_test_type'].str.split(',').str[0]

C:\Users\Bertha\AppData\Local\Temp\ipykernel_6740\3362201690.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['active_ingredients'] = data['molecule_test_type'].str.split(',').str[0]


In [54]:
data['active_ingredients'].value_counts()

active_ingredients
HIV 1/2                                                  1519
Efavirenz                                                1123
Nevirapine                                                877
Lamivudine/Nevirapine/Zidovudine                          707
Lamivudine/Zidovudine                                     689
Lopinavir/Ritonavir                                       633
Lamivudine                                                592
Zidovudine                                                529
Abacavir                                                  453
Tenofovir Disoproxil Fumarate                             320
Lamivudine/Tenofovir Disoproxil Fumarate                  301
Efavirenz/Lamivudine/Tenofovir Disoproxil Fumarate        286
Lamivudine/Nevirapine/Stavudine                           286
Stavudine                                                 284
Didanosine                                                263
Emtricitabine/Tenofovir Disoproxil Fumarate        

In [55]:
# now 52 unique active ingredients
data['active_ingredients'].nunique()

52

In [56]:
data['active_ingredients'].value_counts()

active_ingredients
HIV 1/2                                                  1519
Efavirenz                                                1123
Nevirapine                                                877
Lamivudine/Nevirapine/Zidovudine                          707
Lamivudine/Zidovudine                                     689
Lopinavir/Ritonavir                                       633
Lamivudine                                                592
Zidovudine                                                529
Abacavir                                                  453
Tenofovir Disoproxil Fumarate                             320
Lamivudine/Tenofovir Disoproxil Fumarate                  301
Efavirenz/Lamivudine/Tenofovir Disoproxil Fumarate        286
Lamivudine/Nevirapine/Stavudine                           286
Stavudine                                                 284
Didanosine                                                263
Emtricitabine/Tenofovir Disoproxil Fumarate        

Majority of products are test kits to diagnose HIV 1 and 2, hence HIV1/2. The second most common is Efavirenz, which is a medication to treat those with HIV.

In [57]:
data.isnull().sum()

project_code                       0
pq_num                             0
po___so_num                        0
asn_dn_num                         0
country                            0
managed_by                         0
fulfill_via                        0
vendor_inco_term                   0
shipment_mode                    360
pq_first_sent_to_client_date    5729
po_sent_to_vendor_date          5729
scheduled_delivery_date            0
delivered_to_client_date           0
delivery_recorded_date             0
product_group                      0
sub_classification                 0
vendor                             0
molecule_test_type                 0
brand                              0
dosage                          1734
dosage_form                        0
unit_of_measure_per_pack           0
line_item_quantity                 0
line_item_value                    0
pack_price                         0
unit_price                         0
manufacturing_site                 0
f

## 3. manufacturing_site

88 unique manufacturing sites that partner with USAID and ship commodities and treatments for HIV/AIDS, Malaria

In [58]:
# before alphabetical sorting
data['manufacturing_site'].unique()

array(['Ranbaxy Fine Chemicals LTD', 'Aurobindo Unit III, India',
       'ABBVIE GmbH & Co.KG Wiesbaden', 'Ranbaxy, Paonta Shahib, India',
       'MSD South Granville Australia', "ABBVIE (Abbott) St. P'burg USA",
       'ABBVIE Ludwigshafen Germany', 'Trinity Biotech, Plc',
       'EY Laboratories, USA', 'Cipla, Goa, India', 'BMS Meymac, France',
       'Premier Med. Corp Ltd. India', 'Chembio Diagnostics Sys. Inc.',
       'Inverness Japan', 'Pacific Biotech, Thailand',
       'Standard Diagnostics, Korea', 'GSK Mississauga (Canada)',
       'Gilead(Nycomed) Oranienburg DE', 'Bio-Rad Laboratories',
       'Mylan (formerly Matrix) Nashik', 'Roche Basel', 'GSK Ware (UK)',
       'Novartis Pharma AG, Switzerland', 'BI, Ingelheim, Germany',
       'Not Applicable', 'Ipca Dadra/Nagar Haveli IN', 'MSD, Haarlem, NL',
       'Aspen-OSD, Port Elizabeth, SA', 'ABBVIE (Abbott) Logis. UK',
       'Gland Pharma Ltd Pally Factory', 'GSK Aranda',
       'Hetero Unit III Hyderabad IN', 'ABBVIE (Abbot

In [59]:
manufacturing = data.sort_values('manufacturing_site')

# after alphabetical sorting
manufacturing['manufacturing_site'].unique()

array(['ABBSP', 'ABBVIE (Abbott) France',
       'ABBVIE (Abbott) Japan Co. Ltd.', 'ABBVIE (Abbott) Logis. UK',
       "ABBVIE (Abbott) St. P'burg USA", 'ABBVIE GmbH & Co.KG Wiesbaden',
       'ABBVIE Labs North Chicago US', 'ABBVIE Ludwigshafen Germany',
       'Access BIO, L.C.', 'Alere Medical Co., Ltd.',
       'Aspen-OSD, Port Elizabeth, SA', 'Aurobindo Unit III, India',
       'Aurobindo Unit VII, IN', 'BI, Ingelheim, Germany',
       'BMS Evansville, US', 'BMS Meymac, France',
       'BUNDI INTERNATIONAL DIAGNOSTICS LTD', 'Bio-Rad Laboratories',
       'Boehringer Ing., Koropi, GR', 'Boehringer Ingelheim Roxane US',
       'Bristol-Myers Squibb Anagni IT', 'Chembio Diagnostics Sys. Inc.',
       'Cipla Ltd A-42 MIDC Mahar. IN', 'Cipla, Goa, India',
       'Cipla, Kurkumbh, India', 'Cipla, Patalganga, India',
       'EY Laboratories, USA',
       'Emcure Plot No.P-2, I.T-B.T. Park, Phase II, MIDC, Hinjwadi, Pune, India',
       'GSK Aranda', 'GSK Barnard Castle UK',
       'GSK C

'Emcure Plot No.P-2, I.T-B.T. Park, Phase II, MIDC, Hinjwadi, Pune, India', is a typo that includes the address, should just be renamed 'Emcure Pharmaceuticals Limited, India'

'Weifa A.S., Hausmanngt. 6, P.O. Box 9113 Grønland, 0133, Oslo, Norway'

After sorting we see there are no duplicates, but Ranbaxy similar name

In [60]:
# print rows with  'Ranbaxy per Shasun Pharma' or 'Ranbaxy per Shasun Pharma Ltd'
data.loc[data['manufacturing_site'] == 'Ranbaxy per Shasun Pharma']

project_code     pq_num po___so_num asn_dn_num     country managed_by  \
id                                                                            
82703   113-ZW-T30  FPQ-13508    SO-47540    DN-3825    Zimbabwe   PMO - US   
82782   107-RW-T30  FPQ-14111    SO-48810    DN-4149      Rwanda   PMO - US   
84919   110-ZM-T30  FPQ-13221    SO-47201    DN-3648      Zambia   PMO - US   
84980   110-ZM-T30  FPQ-13221    SO-47207    DN-3941      Zambia   PMO - US   
85572   102-BI-T30  FPQ-15406    SO-50280    DN-4178     Burundi   PMO - US   
85585   102-BI-T30  FPQ-15406    SO-50290    DN-4210     Burundi   PMO - US   
86677   110-ZM-T30  FPQ-13221    SO-47201    DN-3648      Zambia   PMO - US   
86721   111-MZ-T30  FPQ-14226    SO-48870    DN-3903  Mozambique   PMO - US   
86786   114-UG-T30  FPQ-14396    SO-48990    DN-4163      Uganda   PMO - US   

      fulfill_via vendor_inco_term shipment_mode pq_first_sent_to_client_date  \
id                                                                              
82703    From RDC   N/A - From RDC         Truck                          NaT   
82782    From RDC   N/A - From RDC         Truck                          NaT   
84919    From RDC   N/A - From RDC         Truck                          NaT   
84980    From RDC   N/A - From RDC         Truck                          NaT   
85572    From RDC   N/A - From RDC           Air                          NaT   
85585    From RDC   N/A - From RDC           Air                          NaT   
86677    From RDC   N/A - From RDC         Truck                          NaT   
86721    From RDC   N/A - From RDC         Truck                          NaT   
86786    From RDC   N/A - From RDC         Truck                          NaT   

       ... line_item_quantity line_item_value pack_price unit_price  \
id     ...                                                            
82703  ...              76950       356278.50       4.63       0.15   
82782  ...              21868       100374.12       4.59       0.15   
84919  ...              29440       136601.60       4.64       0.15   
84980  ...              11536        54219.20       4.70       0.16   
85572  ...                240         1101.60       4.59       0.15   
85585  ...                240         1101.60       4.59       0.15   
86677  ...              12590        58417.60       4.64       0.15   
86721  ...              38774       179911.36       4.64       0.15   
86786  ...              27365       125605.35       4.59       0.15   

              manufacturing_site first_line_designation  \
id                                                        
82703  Ranbaxy per Shasun Pharma                  False   
82782  Ranbaxy per Shasun Pharma                   True   
84919  Ranbaxy per Shasun Pharma                  False   
84980  Ranbaxy per Shasun Pharma                   True   
85572  Ranbaxy per Shasun Pharma                  False   
85585  Ranbaxy per Shasun Pharma                  False   
86677  Ranbaxy per Shasun Pharma                  False   
86721  Ranbaxy per Shasun Pharma                  False   
86786  Ranbaxy per Shasun Pharma                   True   

              weight_kilograms         freight_cost_usd  \
id                                                        
82703  See DN-3825 (ID#:82702)  See DN-3825 (ID#:82702)   
82782                     4326                  9426.33   
84919  See DN-3648 (ID#:84360)  See DN-3648 (ID#:84360)   
84980                     1046                  1157.78   
85572  See DN-4178 (ID#:84460)  See DN-4178 (ID#:84460)   
85585  See DN-4210 (ID#:83906)  See DN-4210 (ID#:83906)   
86677  See DN-3648 (ID#:84360)  See DN-3648 (ID#:84360)   
86721  See DN-3903 (ID#:86131)  See DN-3903 (ID#:86131)   
86786                     2885                  2564.94   

      line_item_insurance_usd                        active_ingredients  
id                                                                       
82703         

In [61]:
data.loc[data['manufacturing_site'] == 'Ranbaxy per Shasun Pharma Ltd']

project_code     pq_num  po___so_num asn_dn_num   country managed_by  \
id                                                                           
75352   901-CM-T30  FPQ-12837  SCMS-204370  ASN-22180  Cameroon   PMO - US   
84360   110-ZM-T30  FPQ-13221     SO-47201    DN-3648    Zambia   PMO - US   

       fulfill_via vendor_inco_term shipment_mode  \
id                                                  
75352  Direct Drop              EXW           Air   
84360     From RDC   N/A - From RDC         Truck   

      pq_first_sent_to_client_date  ... line_item_quantity line_item_value  \
id                                  ...                                      
75352                   2013-09-25  ...              39960        193406.4   
84360                          NaT  ...              27900        129456.0   

      pack_price unit_price             manufacturing_site  \
id                                                           
75352       4.84       0.16  Ranbaxy per Shasun Pharma Ltd   
84360       4.64       0.15  Ranbaxy per Shasun Pharma Ltd   

      first_line_designation weight_kilograms freight_cost_usd  \
id                                                               
75352                   True             4123         28778.76   
84360                   True            17018         19564.23   

      line_item_insurance_usd                        active_ingredients  
id                                                                       
75352                  239.24  Lamivudine/Tenofovir Disoproxil Fumarate  
84360                  133.08  Lamivudine/Tenofovir Disoproxil Fumarate  

[2 rows x 32 columns]

lines 86677 and 84360 have the same project_code, pq_num, etc. to the same location, but 
line_item_quantity, line_item_value, and pack_price differ
It also says "Ranbaxy per Shasun Pharma"


## 4. weight_kilograms

4688 unique weights

In [62]:
weightkg = data.sort_values('weight_kilograms')
weightkg['weight_kilograms'].value_counts()

weight_kilograms
Weight Captured Separately    1507
2                               29
6                               26
1                               23
5                               20
                              ... 
3723                             1
3728                             1
3732                             1
3733                             1
4873                             1
Name: count, Length: 4688, dtype: int64

In [63]:
path = "C:/Users/Bertha/Documents/SJSU/5_Spring2024/DATA_240-Data_Mining_Analytics/DATA240_Project/DATA240/data/Supply_Chain_Shipment_Pricing_Dataset_20240302.csv"
data1 = pd.read_csv(path) # ID col is treated as the index
data1.head()

id project code            pq # po / so # asn/dn #        country  \
0   1   100-CI-T01  Pre-PQ Process    SCMS-4    ASN-8  Côte d'Ivoire   
1   3   108-VN-T01  Pre-PQ Process   SCMS-13   ASN-85        Vietnam   
2   4   100-CI-T01  Pre-PQ Process   SCMS-20   ASN-14  Côte d'Ivoire   
3  15   108-VN-T01  Pre-PQ Process   SCMS-78   ASN-50        Vietnam   
4  16   108-VN-T01  Pre-PQ Process   SCMS-81   ASN-55        Vietnam   

  managed by  fulfill via vendor inco term shipment mode  ...  \
0   PMO - US  Direct Drop              EXW           Air  ...   
1   PMO - US  Direct Drop              EXW           Air  ...   
2   PMO - US  Direct Drop              FCA           Air  ...   
3   PMO - US  Direct Drop              EXW           Air  ...   
4   PMO - US  Direct Drop              EXW           Air  ...   

  unit of measure (per pack) line item quantity line item value pack price  \
0                         30                 19           551.0      29.00   
1                        240               1000          6200.0       6.20   
2                        100                500         40000.0      80.00   
3                         60              31920        127360.8       3.99   
4                         60              38000        121600.0       3.20   

  unit price             manufacturing site first line designation  \
0       0.97     Ranbaxy Fine Chemicals LTD                   True   
1       0.03      Aurobindo Unit III, India                   True   
2       0.80  ABBVIE GmbH & Co.KG Wiesbaden                   True   
3       0.07  Ranbaxy, Paonta Shahib, India                   True   
4       0.05      Aurobindo Unit III, India                   True   

  weight (kilograms) freight cost (usd) line item insurance (usd)  
0                 13             780.34                       NaN  
1                358             4521.5                       NaN  
2                171            1653.78                       NaN  
3               1855           16007.06                       NaN  
4               7590           45450.08                       NaN  

[5 rows x 33 columns]

In [64]:
data1.shape[1]

33

In [65]:
# Create key & value pair for dictionaries: id & weight and freight
weightid = dict(zip(data1['id'], data1['weight (kilograms)']))
freightid = dict(zip(data1['id'], data1['freight cost (usd)']))

In [66]:
def weight(data1):
    data1 = pd.to_numeric(data1, errors = 'ignore')
    if type(data1) == str:
        try:
            d = data1.split(':')
            e = d[1][:-1]
            return weightid.get(int(e))
        except:
            return (data1)
    else:
        return (data1)

In [67]:
def fright(data1):
    data1 = pd.to_numeric(data1, errors = 'ignore')
    if type(data1) == str:
        try:
            d = data1.split(':')
            e = d[1][:-1]
            return freightid.get(int(e))
        except:
            return (data1)
    else:
        return (data1)

In [68]:
# analysis of traget columns : Freight Cost (USD)
# total no of different values
data1['freight cost (usd)'].value_counts()

freight cost (usd)
Freight Included in Commodity Cost    1442
Invoiced Separately                    239
9736.1                                  36
6147.18                                 27
See DN-304 (ID#:10589)                  16
                                      ... 
12793.7                                  1
35499.58                                 1
1949.64                                  1
569.11                                   1
See DN-4282 (ID#:83919)                  1
Name: count, Length: 6733, dtype: int64

In [69]:
# Replace missing values in Weight & Freight columns 
data1['freight cost (usd)'] = data1['freight cost (usd)'].replace('Freight Included in Commodity Cost', 0)
# do not impute by 0's

In [70]:
data1['weight (kilograms)'] = data1['weight (kilograms)'].apply(get_weight)
data1['freight cost (usd)'] = data1['freight cost (usd)'].apply(get_freight)

NameError: name 'get_weight' is not defined

In [ ]:
data1['weight (kilograms)'] = pd.to_numeric(data1['weight (kilograms)'], errors = 'coerce')
data1['freight cost (usd)'] = pd.to_numeric(data1['freight cost (usd)'], errors = 'coerce')

In [ ]:
data1['weight (kilograms)'] = data1['weight (kilograms)'].fillna(data1['weight (kilograms)'].mean())
data1['freight cost (usd)'] = data1['freight cost (usd)'].fillna(data1['freight cost (usd)'].mean())

In [ ]:
data1.isnull().sum() # no more nulls

In [ ]:
data1['weight (kilograms)'].value_counts()

In [ ]:
data1['freight cost (usd)'].value_counts()

In [ ]:
data1

In [ ]:
data

## 5. freight_cost_usd

6733 unique freight costs

In [ ]:
fcost['weight_kilograms'].nunique()

In [ ]:
fcost = data.sort_values('freight_cost_usd')
fcost['freight_cost_usd'].unique()

In [ ]:
fcost['freight_cost_usd'].nunique()

Followed these tutorials and links:
* https://www.kaggle.com/code/sumeet07/shipping-price-prediction-eda
* https://www.kaggle.com/code/divyeshardeshana/supply-chain-shipment-price-data-analysis/notebook
* https://www.kaggle.com/code/rushikeshkalkar/supply-chain-shipment

We expanded upon these tutorials, drawing insights and relationships of different factors that affect pharmaceutical supply chains in regards to HIV/AIDS and malaria.